Create world headless

BoySanic opened the pull request:
Fixes #2250 
Allows for creating a world headlessly.

IntegratedQuantum commented on build.zig:
```diff
@@ -148,6 +148,11 @@ fn createLaunchConfig() !void {
 			\\    .autoEnterWorld = "",
 			\\    .headlessServer = false,
+			\\    .worldConfig = .{
+			\\            .gamemode = "survival",
```
Please consistently use 4 space indentation

IntegratedQuantum commented on src/main.zig:
```diff
@@ -573,6 +573,10 @@ pub fn main() void { // MARK: main()
 	defer server.terrain.globalDeinit();
 
 	if(headless) {
+		server.world_zig.tryCreateWorld(settings.launchConfig.autoEnterWorld, settings.launchConfig.worldConfig) catch |err| {
```
Doesn't this mean that it recreates the world every time, likely overwriting existing world settings (e.g. game rules) in the process.

Please check if the world exists first.

BoySanic replied:
https://github.com/PixelGuys/Cubyz/blob/8df1b04a29718068e4be1dacf5254ad41b5800fe/src/server/world.zig#L68

I thought it would fail to create the world if it already existed, but I can add some logic to check that it does before attempting to create it.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -30,7 +30,29 @@ pub const WorldSettings = struct {
+			};
+		} else {
+			self = .{
+				.gamemode = .creative,
```
Please use the same defaults everywhere.
Ideally just use the field defaults directly:
```zig
var result: WorldSettings = .{};
result = .{
	.testingMode = zon.get(bool, "testingMode", result.testingMode);
...
};
```

IntegratedQuantum commented on src/main.zig:
```diff
@@ -573,10 +573,19 @@ pub fn main() void { // MARK: main()
-			return;
+		const saveDirectory = std.fs.path.join(stackAllocator.allocator, &.{settings.launchConfig.cubyzDir, "saves", settings.launchConfig.autoEnterWorld}) catch unreachable;
+		defer stackAllocator.free(saveDirectory);
+		var world_found = true;
```
```suggestion
		var worldFound = true;
```

IntegratedQuantum commented on build.zig:
```diff
@@ -148,6 +148,15 @@ fn createLaunchConfig() !void {
 			\\    .autoEnterWorld = "",
 			\\    .headlessServer = false,
+			\\    // This is used for world creation when autoEnterWorld is set and the world doesn't exist yet.
+			\\    // .seed = "" will randomly generate a seed. If you enter a numeric seed, it will be interpreted as such (i.e. "12345").
```
Please add this directly in the line of or next to the seed field.

IntegratedQuantum commented on src/server/world.zig:
```diff
@@ -31,8 +31,42 @@ pub const Settings = struct {
+	files.cubyzDir().dir.access(saveDirectory, .{}) catch |err| {
+		if(err == error.FileNotFound) worldFound = false;
+	};
+	return worldFound;
```
```suggestion
	files.cubyzDir().dir.access(saveDirectory, .{}) catch |err| {
		if(err == error.FileNotFound) return false;
	};
	return true;
```

IntegratedQuantum reviewed (changes requested):
also you got more conflicts

BoySanic commented:
ok this one should be ready to go again

BoySanic commented:
I've changed `fromZon` to not return an error, and added error handling in `loadWorld` instead so we can have it not proceed with loading the world if it discovers the seed is missing.

We could instead give it a new seed to work with, but I think it makes more sense to try to avoid potential corruption.

I also deduplicated the code used in autoEnterWorld on the client and headless for the server.

IntegratedQuantum commented on src/settings.zig:
```diff
@@ -210,8 +210,9 @@ pub const launchConfig = struct {
+	pub var worldCreationSettings: main.server.world_zig.Settings = undefined;
+	pub var worldCreationPreset: []const u8 = "";
 	pub var preferredAuthenticationAlgorithm: main.network.authentication.KeyTypeEnum = .ed25519;
-
```
unintended change

BoySanic commented on src/server/world.zig:
```diff
@@ -62,6 +59,19 @@ pub const Settings = struct {
+pub fn exists(worldPath: []const u8) !bool {
+	const saveDirectory = std.fs.path.join(main.stackAllocator.allocator, &.{"saves", worldPath, "world.zig.zon"}) catch unreachable;
+	defer main.stackAllocator.free(saveDirectory);
+	files.cubyzDir().dir.access(saveDirectory, .{}) catch |err| {
```
I can do that, but we lose out on providing the user an error, since we don't log it in `hasFile` or `hasDir`.

IntegratedQuantum commented:
> I made two new functions that return errors (tryHasDir, tryHasFile)
>
> Making hasFile and hasDir return errors has a huge blast radius

The problem with this approach is that it leads to a bunch of different functions and no one knows which they should call.
And we are just talking about 6 use-cases.

If you are unsure how to handle a case, then just use `catch false` to preserve the status quo behavior. I'll then check in the review if we should do it differently.

IntegratedQuantum commented on src/files.zig:
```diff
@@ -125,14 +125,26 @@ pub const Dir = struct {
+				return false;
+			} else {
+				return err;
+			}
```
Please do this at least a little more compact.
```suggestion
			}
			return err;
```

BoySanic commented on src/server/world.zig:
```diff
@@ -38,16 +38,26 @@ pub const Settings = struct {
 			.allowCheats = zon.get(bool, "allowCheats", defaults.allowCheats),
 			.testingMode = zon.get(bool, "testingMode", defaults.testingMode),
 		};
+		self.chooseSeed(zon.get([]const u8, "seed", ""));
```
I've changed this so that it tries a u64 first. If that fails, it has a fallback to trying chooseSeed for a []const u8.

I also generally cleaned up the implementation a bit

IntegratedQuantum commented on src/files.zig:
```diff
@@ -128,14 +128,24 @@ pub const Dir = struct {
 
-	pub fn hasFile(self: Dir, subPath: []const u8) bool {
-		const file = self.dir.openFile(main.io, subPath, .{}) catch return false;
+	pub fn hasFile(self: Dir, subPath: []const u8) !bool {
```
Sorry for this, but would you mind doing this change in a separate PR first, which also describes why this is needed? I have totally forgot why we decided on doing this, and it's a change that doesn't really belong here and touches many unrelated sites.
